File: NeuroEvolution/flappyBird/sketch.js
const TOTAL_POPULATION = 200;
const GA_ACTIVE = false;
const LOAD_MODEL = true;
let MODEL_LOAD_PATH = "resources/AgentBrain.json";

let activeAgents = [];;
let deadAgents = [];
let blocks = [];
let blockCounter = 0;
let generationCount = 0;

let slider;
let drawButton;
let saveButton;

let loadedBrain;

function preload() {

	if (LOAD_MODEL) {
		loadedBrain = loadJSON(MODEL_LOAD_PATH);
	}

}

function setup() {
	createCanvas(windowWidth * 0.7, windowHeight * 0.5);

	// Make the Interface Stuff
	slider = createSlider(1, 200, 1);
	drawButton = createCheckbox("draw sim", true);
	saveButton = createButton("Save Best Agent");
	saveButton.mousePressed(saveButtonPressed);

	loadAgents();


}

function loadAgents() {
	if (LOAD_MODEL) {

		//let brain = NeuralNetwork.deserialize(loadedModelJSON);
		//let brain = NeuralNetwork.deserialize(loadedBrain);
		let agent = new Agent(loadedBrain, true);
		activeAgents.push(agent);

	} else {
		for (let i = 0; i < TOTAL_POPULATION; i++) {
			activeAgents.push(new Agent(false, false));
		}
	}
}

function draw() {

	let cycles = slider.value();

	for (let c = 0; c < cycles; c++) {
		if (blockCounter % 75 === 0) {
			blocks.push(new block());
		}
		blockCounter++;


		for (let i = 0; i < blocks.length; i++) {

			blocks[i].update();

			if (blocks[i].offscreen()) {
				blocks.splice(i, 1);
				i--;
			}
		}


		// All agents will have the same xPos 
		let globalXPos = activeAgents[0].xPos;

		// Find the closestBloct. because the agets all have the same xPos then 
		// it means the closest block for one will be the closest block for all

		let clostestBlock = null;

		for (let i = 0; i < blocks.length; i++) {
			if (globalXPos < (blocks[i].xPos + blocks[i].width)) {
				clostestBlock = blocks[i];
				break;
			}
		}

		// Go through each agent check if they have hit something and if they have 
		// add them to the deadAgents list. 

		for (let i = 0; i < activeAgents.length; i++) {

			let agent = activeAgents[i];
			agent.think(clostestBlock);
			agent.update(clostestBlock);

			if (clostestBlock.hit(agent)) {
				// Now we penelise the agent based on their average distance from the center of the gaps
				// over the course of its entire life. this is to help distinguish between the performance
				// of agents that all crash into the same block, the ones that were closer to the gap, were
				// on a better track than the ones miles away from it.
				agent.computeFitness();
				deadAgents.push(agent);
				activeAgents.splice(i, 1);
				i--;
			}

		}

		// If everything is dead resart the game with new population.

		if (activeAgents.length === 0) {
			if (GA_ACTIVE) {
				activeAgents = GA.produceNextGeneration(deadAgents);
				generationCount++;
			} else {
				loadAgents();
			}
			resetGame();

			console.log(generationCount)
		}



	}

	if (drawButton.checked()) {
		// Draw Everything

		background(255);

		for (let i = 0; i < blocks.length; i++) {
			blocks[i].show();
		}

		for (let i = 0; i < activeAgents.length; i++) {
			activeAgents[i].show();
		}
	}



}

function resetGame() {
	deadAgents = [];
	blocks = [];
	blockCounter = 0;
}



function saveButtonPressed() {

	if (activeAgents.length > 0) {

		let a = activeAgents[0];
		a.computeFitness();
		let max = a.score;

		let index = 0;

		for (let i = 0; i < activeAgents.length; i++) {

			let currentAgent = activeAgents[i];
			currentAgent.computeFitness();

			if (currentAgent > max) {
				max = currentAgent.score;
				index = i;
			}

		}

		saveJSON(activeAgents[index].brain, 'AgentBrain.json');
	}
}


// function resetGame() {
// 	deadAgents = [];
// 	activeAgents = [];
// 	for (let i = 0; i < TOTAL_POPULATION; i++) {
// 		activeAgents.push(new Agent());
// 	}
// }


// p5 if you define a function in the sketch it will get executed
// The keyPressed() function is called once every time a key is pressed. 
//The keyCode for the key that was pressed is stored in the keyCode variable. 
// function keyPressed() {

// 	if (key === ' ') {
// 		agent.actionUp();
// 	}
// }

File: python_models/src_STM_model/src_pycharm/sketch.py
import numpy as np
from p5 import *
from agent import Agent
from block import Block
from GA import GA

class Simulation:

    def __init__(self):

        self.TOTAL_POPULATION = 400
        self.GA_ACTIVE = True
        self.LOAD_MODEL = False
        self.MODEL_LOAD_PATH = ""
        self.CYCLES = 1

        self.activeAgents = []
        self.deadAgents = []
        self.blocks = []
        self.blockCounter = 0
        self.generationCount = 0

        self.slider = None
        self.drawButton = None
        self.saveButton = None

        self.loadedBrain = None



def setup():

    size(800, 300)
    sim = Simulation()
    load_agents(sim)




def load_agents(sim):

    for i in range(sim.TOTAL_POPULATION):
        sim.activeAgents.append(Agent(initBrain=False, initEmpty=False))



def draw():
    sim = Simulation()
    print(sim.blockCounter)

    for c in range(sim.CYCLES):

        if sim.blockCounter % 75 == 0:
            sim.blocks.append(Block())
        sim.blockCounter = sim.blockCounter + 1

        i = 0
        while i < len(sim.blocks):
            sim.blocks[i].update()

            if sim.blocks[i].offscreen():
                sim.blocks.pop(i)
                i = i - 1

            i = i + 1

        #all agents will have the same xPos
        globalXPos = sim.activeAgents[0].xPos

        # find closestblock. all agents have same x pos and it never changes so the closest block
        # for one will be the closest block for all

        closestblock = None

        for i in range(len(sim.blocks)):
            if globalXPos < sim.blocks[i].xPos + sim.blocks[i].width:
                closestblock = sim.blocks[i]
                break

        # Go through each agent check if they have hot something and if they have add them to deadAgents list
        i = 0
        while i < len(sim.activeAgents):

            agent = sim.activeAgents[i]
            agent.think(closestblock)
            agent.update(closestblock)

            if closestblock.hit(agent):
                # Now we penelise the agent based on their average distance from the center of the gaps
                # over the course of its entire life. this is to help distinguish between performance
                # of agents that all crash into the same block, the ones that were closer to the gap,
                # were on a better track than the ones away from it

                agent.computeFitness()
                sim.deadAgents.append(agent)
                sim.activeAgents.pop(i)
                i = i - 1

            i = i + 1


        # if everything is dead restart the game

        if len(sim.activeAgents) == 0:
            sim.activeAgents = GA.produceNextGeneration(sim.deadAgents)
            sim.generationCount = sim.generationCount + 1


        # draw everything
        background(255)

        for i in range(len(sim.blocks)):
            sim.blocks[i].show()

        for i in range(len(sim.activeAgents)):
            sim.activeAgents[i].show()


run()
